components/server/src/util/request-context.ts

```diff
@@ -7,32 +7,161 @@
 import { AsyncLocalStorage } from "node:async_hooks";
 import { performance } from "node:perf_hooks";
 import { v4 } from "uuid";
+import { SubjectId } from "../auth/subject-id";
+import { ApplicationError, ErrorCodes } from "@gitpod/gitpod-protocol/lib/messaging/error";
 
+/**
+ * ReqeuestContext is the context that all our request-handling code runs in.
+ * All code has access to the contained fields by using the exported "ctx...()" functions below.
+ *
+ * It's meant to be the host all concerns we have for a request. For now, those are:
+ *  - authorization (via subjectId)
+ *  - caching (via cache, ctxSetCache, ctxGetCache)
+ *
+ * It's meant to be nestable, so that we can run code in a child context with different properties.
+ * The only example we have for now is "runWithSubjectId", which executes the child context with different authorization.
+ * @see runWithSubjectId
+ */
 export interface RequestContext {
-    contextId?: string;
-    contextKind?: string;
-    contextTimeMs?: number;
+    /**
+     * Unique, artificial ID for this request.
+     */
+    readonly requestId: string;
+
+    /**
+     * A request kind e.g. "job", "http" or "grpc".
+     */
+    readonly requestKind: string;
+
+    /**
+     * A name. Specific values depends on requestKind.
+     */
+    readonly requestMethod: string;
+
+    /**
+     * Propagate cancellation through the handler chain.
+     */
+    readonly signal: AbortSignal;
+
+    /**
+     * The UNIX timestamp in milliseconds when request processing started.
+     */
+    readonly startTime: number;
+
+    /**
+     * A cache for request-scoped data.
+     */
+    readonly cache: { [key: string]: any };
+
+    /**
+     * The trace ID for this request. This is used to correlate log messages and other events.
+     */
+    readonly traceId?: string;
+
+    /**
+     * The SubjectId this request is authenticated with.
+     */
+    readonly subjectId?: SubjectId;
 }
 
 const asyncLocalStorage = new AsyncLocalStorage<RequestContext>();
+
+export function ctxGet(): RequestContext {
+    const ctx = asyncLocalStorage.getStore();
+    if (!ctx) {
+        throw new Error("ctxGet: No request context available");
+    }
+    return ctx;
+}
+
+export function ctxTryGet(): RequestContext | undefined {
+    return asyncLocalStorage.getStore() || undefined;
+}
+
+/**
+ * @returns the SubjectId this request is authenticated with, or undefined if it isn't
+ */
+export function ctxTrySubjectId(): SubjectId | undefined {
+    const ctx = ctxTryGet();
+    return ctx?.subjectId;
+}
+
 /**
- * !!! Only to be used by selected internal code !!!
+ * @throws 500/INTERNAL_SERVER_ERROR if there is no userId
+ * @returns The userId associated with the current request.
+ */
+export function ctxUserId(): string {
+    const userId = ctxGet()?.subjectId?.userId();
+    if (!userId) {
+        throw new ApplicationError(ErrorCodes.INTERNAL_SERVER_ERROR, "No userId available");
```

Comment:
I don't think so: In the future, it might very well be that there are authorization methods that don't return a userId here. If we still rely on this then, it's application issue.
This method is primarily useful for filling the "first argument userId: string" until we remove it.

Reply:
ok